### PR TITLE
daemon: allow setting OCI annotations on containers

### PR DESCRIPTION
**Pull request by corhere**
- Relates to #45030
- Closes #39142
- Closes #37289
- Closes #36181
- Relates to #37262

**- What I did**

Allow API clients to set and retrieve OCI annotations on a container.

**- How I did it**

Added a field to `HostConfig` and plumbed it into the OCI spec generation code.

**- How to verify it**
Create a container with annotations and start it. Inspect the running container using `ctr` and verify that the annotations are listed in the container spec.

A new integration test verifies that annotations can be round-tripped through the Engine API.

**- Description for the changelog**

- Annotations—arbitrary non-identifying metadata—can now be attached to containers on creation which will be passed to the runtime when the container is started.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Will these labels be committed in the image on `docker commit` ? (also see https://github.com/moby/moby/pull/42037) And does this mean that an image containing these labels will set them at runtime on the container?

**Comment by corhere** [comment]
> Will these labels be committed in the image on `docker commit` ? (also see #42037)

Yes, until the aforementioned PR is merged.

> And does this mean that an image containing these labels will set them at runtime on the container?

Yes. This could arguably be a feature.

**Comment by thaJeztah** [comment]
> Yes. This could arguably be a feature.

We should probably look closely what's desirable in that case. I recall there was some prior discussion on passing through labels as annotations. I must admit I'm not fully aware how they're used / what they're able to control, but ISTR there were some concerns that had to be looked into (e.g. a malicious image could set a label and do other things than expected).

I guess to some extent that's not "our" issue (as we don't consume those annotations?).

It does somewhat bring back the issue though that we don't have a concept of "runtime" labels and "non-runtime" (image labels are runtime labels in our case, or at least inherited directly, and in case of "commit", baked into the image).

Probably also need to be looked at what the expectations are for `docker container update`, as there's been quite some requests on being able to update labels; prior discussions stranded on that one, but if that would be implemented, that may have an effect on the feature in this PR.

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
> > And does this mean that an image containing these labels will set them at runtime on the container?
>
> Yes. This could arguably be a feature.

This looks like an attack vector.

Specifying annotations should require a new API/CLI like 
`docker run --annotation=com.example.foo=bar`.
(similar to `podman run --annotation`)

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
SGTM

But need to update the API history doc and openapi yaml.

File: api/types/container/hostconfig.go
Comment (AkihiroSuda) [inline api/types/container/hostconfig.go:390]: Why `json:",omitempty"` ?

File: api/server/router/container/container_routes.go
Comment (corhere) [inline api/server/router/container/container_routes.go:569]: I'm following past precedent by ignoring the field on create for older API versions but unconditionally including it in the `/containers/{}/json` response for all API versions.

**Review by neersighted — APPROVED** [review]

**Review by vvoland — APPROVED** [review]

**Review by AkihiroSuda — APPROVED** [review]

File: api/types/container/hostconfig.go
Comment (thaJeztah) [inline api/types/container/hostconfig.go:390]: Before we merge; quick check on the naming. Wondering if `OCI` prefix is needed in our API here. I get where we're coming from; mostly thinking along the lines of "(almost) every option we have in our API leads to `<something>` OCI spec.

So wondering if we should keep the abstraction - call it `Annotation`, and keep the fact that that (currently) gets passed on to the OCI to the description only.

**Review by thaJeztah — CHANGES_REQUESTED** [review]
("requesting changes" just to have a quick check per my comment - changes otherwise look good to me)

File: api/types/container/hostconfig.go
Comment (corhere) [reply api/types/container/hostconfig.go:390]: I'm coming around to dropping the `OCI` prefix and adopting essentially [the same semantics as Kubernetes `Annotations`.](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) And by that I mean documenting that Engine API clients can also leverage annotations to attach and consume metadata on containers for their own purposes. Perhaps we should also follow their lead and [enforce a particular syntax for annotation keys](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set) so that the namespace of an annotation is unambiguous. We could always loosen validation later but can't make it more restrictive.

Comment (thaJeztah) [reply api/types/container/hostconfig.go:390]: Let's discuss / brainstorm in the call later Today (I also need to get my head around all of that). Quick comment;

- Yes, I think Annotations could be useful for "our side" as well
- I like namespaces / prefixes as they can definitely help prevent ambiguity (I think I may actually have been the one originally getting them into the "labels" definition 🤔 )
- ^^^ w.r.t. namespaces: I always keep in the back of my mind that `API !== UX`; namespaces _are_ awkward to use, but are (in most cases) not meant for humans. So if we decide to use namespaces, we can still make the UX "different" (`docker run --foo=bar` doing `docker run --annotation=awkwardly-long-namespace.option=bar` behind the scenes)

Comment (corhere) [reply api/types/container/hostconfig.go:390]: What I like about how K8s goes about namespacing annotations is that it makes namespace prefixes optional and [reserves the no-namespace "namespace" for users.](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set:~:text=If%20the%20prefix%20is%20omitted%2C%20the%20annotation%20Key%20is%20presumed%20to%20be%20private%20to%20the%20user.)

Comment (thaJeztah) [reply api/types/container/hostconfig.go:390]: Yup, that was part of my original suggestion for labels; no-namespace is for users; don't use them for automation as they can easily conflict. For automation / tools, always use a prefix

Comment (thaJeztah) [reply api/types/container/hostconfig.go:390]: For those interested; here's the PR with all the fun (well, there's been multiple before that, but that's the one that made it, and that one is already 200+ comments 🤣)

- https://github.com/moby/moby/pull/9882

Comment (corhere) [reply api/types/container/hostconfig.go:390]: Decisions from today's maintainer call:
- Drop the `OCI` prefix from the field name
- Any annotation which is valid per the OCI Runtime spec is permitted
  - Keys can be any JSON string except for the empty string
  - Values can be any JSON string, including the empty string
  - Keys and values can be arbitrarily long

File: daemon/container.go
Comment (thaJeztah) [inline daemon/container.go:310]: I had a look if we should be using an `errdefs.InvalidParameter()` here (and noticed other errors didn't), but it looks like we wrap any errors returned here on the call sites, such as https://github.com/moby/moby/blob/2f0e308c7d4cb59f33664b763fa95542dee3c024/daemon/create.go#L65-L68


So in either case; it's good (currently) to not use an errdefs, but we _could_ (should?) look at these at some points to assign error-types in these locations.

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
@cpuguy83 LGTY? (I see there's a "change requested" from you, and didn't want to dismiss it)

Screenshot 2023-02-24 at 14 56 38

**Review by cpuguy83 — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Opened a PR for the CLI:
- https://github.com/docker/cli/pull/4156

**Pull request merged**
